src/sst_iter.rs

```diff
@@ -206,89 +281,84 @@ impl<'a, H: AsRef<SsTableHandle>> SstIterator<'a, H> {
         self.next_block_idx_to_fetch = num_blocks;
-        self.current_iter = None
+        self.state.stop();
     }
 }
 
-impl<'a, H: AsRef<SsTableHandle>> KeyValueIterator for SstIterator<'a, H> {
+impl KeyValueIterator for SstIterator<'_> {
     async fn next_entry(&mut self) -> Result<Option<RowEntry>, SlateDBError> {
-        loop {
-            let current_iter = if let Some(current_iter) = self.current_iter.as_mut() {
-                current_iter
-            } else if let Some(next_iter) = self.next_iter().await? {
-                self.current_iter.insert(next_iter)
+        while !self.state.is_finished() {
+            let next_entry = if let Some(iter) = self.state.current_iter.as_mut() {
+                iter.next_entry().await?
             } else {
-                return Ok(None);
+                None
             };
 
-            let kv = current_iter.next_entry().await?;
-            match kv {
-                Some(kv) if self.range.contains(&kv.key) => return Ok(Some(kv)),
-                Some(kv) => match self.range.end_bound() {
-                    Unbounded => continue,
-                    Included(end_key) | Excluded(end_key) => {
-                        if kv.key > end_key {
-                            self.end_iteration();
-                            return Ok(None);
-                        } else {
-                            continue;
-                        }
+            match next_entry {
+                Some(kv) => {
+                    if self.view.contains(&kv.key) {
+                        return Ok(Some(kv));
+                    } else if self.view.key_exceeds(&kv.key) {
+                        self.stop()
                     }
-                },
-                None => {
-                    self.current_iter = None;
-                    // We have exhausted the current block, but not necessarily the entire SST,
-                    // so we fall back to the top to check if we have more blocks to read.
                 }
+                None => self.advance_block().await?,
             }
         }
+        Ok(None)
     }
 }
 
-impl<'a, H: AsRef<SsTableHandle>> SeekToKey for SstIterator<'a, H> {
+impl SeekToKey for SstIterator<'_> {
     async fn seek(&mut self, next_key: &[u8]) -> Result<(), SlateDBError> {
-        loop {
-            let current_iter = if let Some(current_iter) = self.current_iter.as_mut() {
-                current_iter
-            } else if let Some(next_iter) = self.next_iter().await? {
-                self.current_iter.insert(next_iter)
-            } else {
-                return Ok(());
-            };
-            current_iter.seek(next_key).await?;
-            if current_iter.is_empty() {
-                self.current_iter = None;
-            } else {
-                return Ok(());
+        if !self.view.contains(next_key) {
+            return Err(SlateDBError::InvalidArgument {
+                msg: "FIXME".to_string(),
+            });
+        }
```

Comment:
why not advance to the end and then return an empty iterator? (or is that what you intended to do with the FIXME?

Reply:
I have been debating this. For `DbIterator`, we treat it as an error to seek to a key which is not in the scan range. We should do the same here to be consistent, or change both behaviors to what you suggest. I guess it seemed more likely to me that seeking to a key outside the range is probably a logical error indicating a bug. What do you think?

Reply:
For now I've just improved the error message, which is what I intended with the FIXME.

Reply:
Fair enough, I guess it's also possible that `next_key` is lower than the bounds of the iterator, in which case naturally we would want to seek to the beginning.

I'm happy to keep it consistent and we can revisit this easily if it causes any unexpected errors.